win: add tests for windows blk_pool

peluse opened the pull request:
Includes changing use of fchmod to chmod in general to simplfy
the Windows usage and also implementing basic file permissions
support for Windows in the tests, either RW or RO. A few other
small tweaks in the basic framework (cleanup) discovered while
adding these tests. Any poolset sets use a dummy match file for
now and are marked XXX to be updated once we've figured out
Windows poolset support.

Also, Win doesn't care about case and some of the new .PS1 files
were upper and some lower. Standardized on upper and updated
the filter in RUNTESTS to ignore *.PS1 instead of *.ps1.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/862)
<!-- Reviewable:end -->

marcinslusarz commented:
Review status: 0 of 78 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/common/set.c, line 303 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJ4tTMw-hsNB-pRDR3U:-KJ4tTMw-hsNB-pRDR3V:1976172315) ([raw file](https://github.com/pmem/nvml/blob/4e535d9dd515ff62c444a1718f4a122562a15775/src/common/set.c#L303)):*
> ```C
> 
> 			if (chmod(part->path, mode)) {
> 				ERR("!fchmod %u/%u/%s", r, p,
> ```

s/fchmod/chmod/ here too

---

*[src/test/blk_nblock/TEST2w.PS1, line 59 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJ4tebTN-6MoHBJrUss:-KJ4tebTN-6MoHBJrUst:-1391716755) ([raw file](https://github.com/pmem/nvml/blob/4e535d9dd515ff62c444a1718f4a122562a15775/src/test/blk_nblock/TEST2w.PS1#L59)):*
> ```
> create_poolset $DIR\testset1 1024:$DIR\testfile11:x
> # XXX need to figure out issues with mapping poolsets with windows
> echo "XXX" > out2w.log
> ```

debugging leftover?

---

*[src/test/blk_nblock/TEST2w.PS1, line 64 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJ4tigHQ70JFEQdA-52:-KJ4tigIlcnjSACUKwug:-1637653725) ([raw file](https://github.com/pmem/nvml/blob/4e535d9dd515ff62c444a1718f4a122562a15775/src/test/blk_nblock/TEST2w.PS1#L64)):*
> ```
> 
> #expect_normal_exit ..\..\x64\debug\blk_nblock$Env:EXESUFFIX `
> #    512:$DIR\testset1 `
> ```

unrelated change

---

*[src/test/unittest/unittest.h, line 127 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJ4u5BQjOV82tPwYPSZ:-KJ4u5BQjOV82tPwYPS_:764276148) ([raw file](https://github.com/pmem/nvml/blob/4e535d9dd515ff62c444a1718f4a122562a15775/src/test/unittest/unittest.h#L127)):*
> ```objc
> typedef struct _stat64 ut_util_stat_t;
> #define ut_util_fstat	_fstat64
> inline int ut_util_stat(const char *path,
> ```

+static

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/862)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: 0 of 78 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/common/set.c, line 303 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJ4tTMw-hsNB-pRDR3U:-KJ5wVInFZ94iGLrcE70:978157159) ([raw file](https://github.com/pmem/nvml/blob/4e535d9dd515ff62c444a1718f4a122562a15775/src/common/set.c#L303)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> s/fchmod/chmod/ here too
</details>
oh yeah, didn't catch the comment.  thanks

---

*[src/test/blk_nblock/TEST2w.PS1, line 59 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJ4tebTN-6MoHBJrUss:-KJ5wZR_9697LZ-UvPt1:-2124835063) ([raw file](https://github.com/pmem/nvml/blob/4e535d9dd515ff62c444a1718f4a122562a15775/src/test/blk_nblock/TEST2w.PS1#L59)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> debugging leftover?
</details>
no, we have a half dozen or so tests marked like this.  We have a todo item in our tracker (Trello) to resolve poolset issues but are not holding up all these other tests getting in place

---

*[src/test/blk_nblock/TEST2w.PS1, line 64 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJ4tigHQ70JFEQdA-52:-KJ5wl_c75bkc-Xv0_5G:1098577037) ([raw file](https://github.com/pmem/nvml/blob/4e535d9dd515ff62c444a1718f4a122562a15775/src/test/blk_nblock/TEST2w.PS1#L64)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> unrelated change
</details>
actually it is related, as I added more tests that became 'placeholders' until the poolset work is done I decided to make them all consistent wrt how they are essentially stubbed out.  This one wasn't done the same way as the new ones so I went back and changed it 

---

*[src/test/unittest/unittest.h, line 127 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJ4u5BQjOV82tPwYPSZ:-KJ5x9wuHOa1NyfRjHW5:5173665) ([raw file](https://github.com/pmem/nvml/blob/4e535d9dd515ff62c444a1718f4a122562a15775/src/test/unittest/unittest.h#L127)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> +static
</details>
OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/862)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Squash your commits, please.

<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> #### win: add tests for windows blk\_pool
> 
> Includes changing use of fchmod to chmod in general to simplfy
> 
> the Windows usage and also implementing basic file permissions
> 
> support for Windows in the tests, either RW or RO. A few other
> 
> small tweaks in the basic framework (cleanup) discovered while
> 
> adding these tests. Any poolset sets use a dummy match file for
> 
> now and are marked XXX to be updated once we've figured out
> 
> Windows poolset support.
> 
> 
> 
> Also, Win doesn't care about case and some of the new .PS1 files
> 
> were upper and some lower. Standardized on upper and updated
> 
> the filter in RUNTESTS to ignore *.PS1 instead of *.ps1.
> 
> 
> 
</details>


---

Reviewed 89 of 91 files at r1, 6 of 6 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/test/blk_pool/TEST0w.PS1, line 67 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJCAbWBzTno0q9WmGte:-KJCAbWBzTno0q9WmGtf:906788667) ([raw file](https://github.com/pmem/nvml/blob/071478a281677947dfba9519e0ee9f7773ba80d6/src/test/blk_pool/TEST0w.PS1#L67)):*
> ```
> #
> expect_normal_exit ..\..\x64\debug\blk_pool$Env:EXESUFFIX `
>     c $DIR\testfile 4096 20 0640
> ```

Frankly, I think we could simply pass `0600` as `mode`, both on Linux and Windows. This way, the output would also be the same and there would be no need to have separate *.match files for each OS.

---

*[src/test/blk_pool/TEST4.PS1, line 58 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJCBr5i_9b-W-wXJOqX:-KJCBr5i_9b-W-wXJOqY:-67950024) ([raw file](https://github.com/pmem/nvml/blob/071478a281677947dfba9519e0ee9f7773ba80d6/src/test/blk_pool/TEST4.PS1#L58)):*
> ```
> setup
> 
> echo $null >> $DIR\testfile
> ```

Perhaps, it would be a good idea to define functions like "chmod", "touch", "export", etc. in unittest.ps1 ? A step towards uniform (identical) TEST* scripts on Linux and Windows. 

Of course, this is a subject for a separate patch.

---

*[src/test/unittest/unittest.h, line 131 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJCFjONc4aqy50NEWOv:-KJCFjONc4aqy50NEWOw:-1686732558) ([raw file](https://github.com/pmem/nvml/blob/071478a281677947dfba9519e0ee9f7773ba80d6/src/test/unittest/unittest.h#L131)):*
> ```objc
> 	int retVal = _stat64(path, st_bufp);
> 	/* clear unused bits to avoid confusion */
> 	st_bufp->st_mode &= 0600;
> ```

Is this really needed? I mean, with this change you cannot use STAT/ut_util_stat to check file type (file/dir/pipe/...). Maybe `... &= 0600 | S_IFMT;`?
OTH, bits 0..5 should be zero anyway, so IMHO there's no need to clear'em out. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/862#-:-KJCE7tEsRg3oWPJmOeZ:2004949701)*
<!-- Sent from Reviewable.io -->

peluse commented:
I'll try.. wish me luck :)

<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Squash your commits, please.
</details>


---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/common/set.c, line 303 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJ4tTMw-hsNB-pRDR3U:-KJ6MuPozzliJem7ucHB:-499146009) ([raw file](https://github.com/pmem/nvml/blob/4e535d9dd515ff62c444a1718f4a122562a15775/src/common/set.c#L303)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> oh yeah, didn't catch the comment.  thanks
</details>
Done.

---

*[src/test/blk_pool/TEST0w.PS1, line 67 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJCAbWBzTno0q9WmGte:-KJCJ5mGXvNkEUxd2Dje:1226063335) ([raw file](https://github.com/pmem/nvml/blob/071478a281677947dfba9519e0ee9f7773ba80d6/src/test/blk_pool/TEST0w.PS1#L67)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Frankly, I think we could simply pass `0600` as `mode`, both on Linux and Windows. This way, the output would also be the same and there would be no need to have separate *.match files for each OS.
</details>
If you think that's OK sure! I'll make the change

---

*[src/test/blk_pool/TEST4.PS1, line 58 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJCBr5i_9b-W-wXJOqX:-KJCJBIUz5h6PSp_8gBx:-438404838) ([raw file](https://github.com/pmem/nvml/blob/071478a281677947dfba9519e0ee9f7773ba80d6/src/test/blk_pool/TEST4.PS1#L58)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Perhaps, it would be a good idea to define functions like "chmod", "touch", "export", etc. in unittest.ps1 ? A step towards uniform (identical) TEST* scripts on Linux and Windows. 
> 
> Of course, this is a subject for a separate patch.
</details>
yeah, quite possibly. Of course we may be in a position to use native bash in server 2016 by the time we get all these ported to PS :)

---

*[src/test/unittest/unittest.h, line 127 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJ4u5BQjOV82tPwYPSZ:-KJ6Nk9Bv9JD510eDmsN:-499146009) ([raw file](https://github.com/pmem/nvml/blob/4e535d9dd515ff62c444a1718f4a122562a15775/src/test/unittest/unittest.h#L127)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> OK
</details>
Done.

---

*[src/test/unittest/unittest.h, line 131 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/862#-KJCFjONc4aqy50NEWOv:-KJCJcixgc1JzSYuxHIr:-628217234) ([raw file](https://github.com/pmem/nvml/blob/071478a281677947dfba9519e0ee9f7773ba80d6/src/test/unittest/unittest.h#L131)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Is this really needed? I mean, with this change you cannot use STAT/ut_util_stat to check file type (file/dir/pipe/...). Maybe `... &= 0600 | S_IFMT;`?
> OTH, bits 0..5 should be zero anyway, so IMHO there's no need to clear'em out. 
</details>
so they're not zero, and I couldn't find defns for at last the first byte.  Not sure what's up with that.  I'll take another look though

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/862#-:-KJCJ-iIHRvJMD3E3X5b:-656320435)*
<!-- Sent from Reviewable.io -->
